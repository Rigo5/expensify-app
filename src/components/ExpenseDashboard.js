import React from 'react';
import ExpenseList from './expensesList';
import ExpenseListFilter from './ExpensesListFilter';

const ExpenseDashboard = () => {
    return (
        <div>
            <p>la merda di cane puzza</p>
            <ExpenseListFilter />
            <ExpenseList />
        </div>
    );
}

export default ExpenseDashboard;