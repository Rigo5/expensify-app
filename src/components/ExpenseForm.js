import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';



class ExpenseForm extends React.Component {
    
    constructor(props){
        super(props); 
        console.log(props);
        this.state = {
            description: props.expense ? props.expense.description : '',
            note: props.expense ? props.expense.note : "Notes which are optional",
            amount: props.expense ? props.expense.amount.toString() : '',
            createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
            calendarFocused: false,
            error: ''
        };
    }

    onDescriptionChange = (e) => {
        const description = e.target.value;
        this.setState((prevState) => {
            return {
                description: description
            }
        })
    }

    onNoteChange = (e) => {
        //se non la metto in una variabile e lo uso direttamente nel set state va in erroe e devo usare e.persist()
        const note = e.target.value;
        this.setState((prevState) => ({
            note: note
        }))
    }

    onAmountChange = (e) => {
        const amount = e.target.value;
        //se matcia formato 122331.12 viene aggiornato altrmimenti sticazzi
        if (amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
            this.setState(() => ({ amount: amount }))
        }
    }

    onDateChange = (createAt) => {
        if (createAt) {
            this.setState(() => ({
                createdAt: createAt
            }))
        }
    }

    onFocusChange = ({ focused }) => {
        this.setState(() => ({
            calendarFocused: focused
        }))
    }

    onSubmit = (e) => {
        e.preventDefault();

        if (!this.state.description || !this.state.amount) {
            //please description and amount 
            this.setState(() => ({
                error: 'Please set description and amount'
            }))
        } else {
            if (this.state.error) {
                this.setState(() => ({
                    error: ''
                }))
            }
            console.log(this.state.createdAt.valueOf());
            //clear the error
            this.props.onSubmit({
                description : this.state.description,
                amount : parseFloat(this.state.amount),
                createdAt : this.state.createdAt.valueOf(),
                note : this.state.note
            })
        }
    }
    render() {
        return (
            <div>
                {this.state.error ? <p>Fuck you self</p> : null}
                <form onSubmit={this.onSubmit}>
                    <input type="text"
                        placeholder="Description"
                        value={this.state.description}
                        onChange={this.onDescriptionChange}
                        autoFocus />
                    <input
                        type="text"
                        placeholder="Amount"
                        value={this.state.amount}
                        onChange={this.onAmountChange} />
                    <SingleDatePicker
                        date={this.state.createdAt}
                        onDateChange={this.onDateChange}
                        focused={this.state.calendarFocused}
                        onFocusChange={this.onFocusChange}
                        numberOfMonths={1}
                        isOutsideRange={(day) => false}
                    />
                    <textarea placeholder={this.state.note} onChange={this.onNoteChange} />
                    {this.state.amount ? <button>Edit Expense</button>: <button>Create Expense</button>}
                </form>
            </div>
        );
    }
}

export default ExpenseForm;