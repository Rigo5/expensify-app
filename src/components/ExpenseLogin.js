import React from 'react';


class ExpenseLogin extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            username: '',
            password: ''
        }
    }

    onUsernameChange = (e) => {
        const value = e.target.value;
        
        this.setState(() => ({
            username : value
        }))
    }

    onPasswordChange = (e) => {
        const value = e.target.value;
        
        this.setState(() =>( {
            password : value
        }))
    }

    onSubmit = (e) => {
        e.preventDefault(); 
    }
    render() {
        return (<div>
            <p>Logga per accedere all'applicazione</p>
            <form onSubmit={this.onSubmit}>
                <input
                    type='text'
                    placeholder='Username'
                    value={this.state.username}
                    onChange={this.onUsernameChange}
                />
                <input
                    type='password'
                    placeholder='Password'
                    value={this.state.password}
                    onChange={this.onPasswordChange}
                />
                <button>Login</button>
            </form>
        </div>
        );
    }
}

export { ExpenseLogin as default };