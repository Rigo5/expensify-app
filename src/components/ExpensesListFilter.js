import React from 'react';
import { connect } from 'react-redux';
import { setTextFilter, sortByAmount, sortByDate, setEndDate, setStartDate } from '../actions/filters';
import { DateRangePicker } from 'react-dates';


export class ExpenseListFilter extends React.Component {
    state = {
        focusedInput: null
    }

    onInputChange = (e) => {
        this.props.onInputChange(e.target.value);
    }
 
    onSelectChange = (e) => {
        e.target.value === 'date' ? this.props.byDate() : this.props.byAmount();
    }

    onDatesChange = ({ startDate, endDate }) => {
        this.props.endDate(endDate);
        this.props.startDate(startDate);
    }

    onFocusChange = ({ calendarFocused }) => {
        this.setState(() => ({
            calendarFocused: calendarFocused
        }))
    }

    render() {
        return (
            <div>
                <input type="text" value={this.props.filters.text} onChange={this.onInputChange} />
                <select onChange={this.onSelectChange}>
                    <option value="date">Date</option>
                    <option value="amount">Amount</option>
                </select>

                <DateRangePicker
                    startDate={this.props.filters.startDate}
                    startDateId="1"
                    endDate={this.props.filters.endDate}
                    endDateId="2"
                    onDatesChange={this.onDatesChange}
                    focusedInput={this.state.focusedInput} 
                    onFocusChange={focusedInput => this.setState({ focusedInput })}
                    isOutsideRange={(day) => false}
                />
            </div>
        );
    }
}

const mapDispatchToProps = (dispatch) => ({
    onInputChange : (value) => dispatch(setTextFilter(value)),
    byDate : () => dispatch(sortByDate()),
    byAmount : () => dispatch(sortByAmount()),
    endDate : (endDate) => dispatch(setEndDate(endDate)),
    startDate : (startDate) => dispatch(setStartDate(startDate))
})

const mapStateToProps = (state) => {
    return {
        filters: state.filters
    };
}

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilter);

