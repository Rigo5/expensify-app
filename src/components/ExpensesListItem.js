import React from 'react';
import { connect } from 'react-redux';
import { removeExpense } from '../actions/expenses';
import { Link } from 'react-router-dom';

export const ExpensesListItem = (props) => (
    <div>
        <Link to={'/edit/' + props.id}>{props.description}</Link>
        <p>{props.amount} - {props.createdAt}</p>
        <button onClick={(e) => {
            props.onClick({ id: props.id });
        }}>Remove</button>
    </div>
)

const mapDispatchToProps = (dispatch) => ({
    onClick : (id) => removeExpense({ id : id})
});


export default connect(undefined, mapDispatchToProps)(ExpensesListItem);