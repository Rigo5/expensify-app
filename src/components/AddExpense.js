import React from 'react';
import ExpenseForm from './ExpenseForm';
import { connect } from 'react-redux';
import { addExpense, editExpense } from '../actions/expenses';


/* un problema qua per il testing è dato da : 
    props.dispatch(addExpense({...expense}))
    props dispatch sarà messa come spy, il problema è che come arg
    ha una referenza a una funzione di un modulo e non è passata come prop
    redux per fortuna ci viene in contro e ci permette di definire come per lo stato
    una funzione che mappa i vari dispatch come props
    il codice di prima diventa props.onSubmit(expense); che è più testabile
    in qunato abbiamo astratto il dispatch. 
    Per evitare inline functions sarebbe meglio portare stateless function a componente
    FAAREMO 
    */
export const AddExpense = (props) => {
    return (
        <div>
            <p>aggiungi spese</p>
            <ExpenseForm
                onSubmit={(expense) => {
                    props.onSubmit(expense);
                    props.history.push('/');
                }}

            />
        </div>
    );
}

const mapDispatchToProps = (dispatch) => ({
    onSubmit : (expense) => dispatch(addExpense(expense))
})

export default connect(undefined, mapDispatchToProps)(AddExpense);