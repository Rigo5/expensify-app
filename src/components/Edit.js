import React from 'react';
import ExpenseForm from './ExpenseForm';
import { connect } from 'react-redux';
import { editExpense } from '../actions/expenses';



export class Edit extends React.Component {


    onSubmit = (expense) => {
        console.log('edited');
        this.props.onSubmit(this.props.expense.id, expense);
        this.props.history.push('/');
    }

    render() {
        return (
            <div>
                <p>Editing l'oggetto: {this.props.expense.id}</p>
                <ExpenseForm expense={this.props.expense} onSubmit={this.onSubmit} />
            </div>);
    }
}

const mapStateToProps = (state, props) => {
    return {
        expense: state.expenses.find((expense) => expense.id === props.match.params.id)
    }
};

const mapDispatchToProps = (dispatch) => ({
    onSubmit : (id, expense) => dispatch(editExpense(id, expense))
})

export default connect(mapStateToProps, mapDispatchToProps)(Edit);