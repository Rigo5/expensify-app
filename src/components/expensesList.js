import React from 'react';
//connect permette di getState e dispatch azioni 
import { connect } from 'react-redux';
import ExpensesListItem from './ExpensesListItem';
import selectExpenses from '../selectors/expenses';

export const ExpenseList = (props) => (
    <div>
        {props.expenses.length === 0 ? (
            <p>No expenses</p>
        ) : (
            props.expenses.map((expense) => (
                <ExpensesListItem key={expense.id} {...expense} />
            ))
        )}
    </div>

);

/* il primo argomento di coonect è un afunzione che specifica che informazioni dello stato devono essere passate
al componente che viene vrappato, il secondo argomento è il componente da wrappare. Gli argomenti passati saranno del tipo
key value paair. Ogni volta che lo store cambia questo viene refreshato in modo da avere i dati sempre freschi*/
const mapStateToProps = (state) => {
    return {
        expenses: selectExpenses(state.expenses, state.filters)
    };
};


export default connect(mapStateToProps)(ExpenseList);
