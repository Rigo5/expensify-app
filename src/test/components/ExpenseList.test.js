import { ExpenseList } from "../../components/expensesList";
import React from 'react';
import { shallow } from 'enzyme';
import moment from "moment";


const testData = [
    {
        id: '1',
        description: 'merda',
        amount : 10,
        note: 'merdamerda',
        createdAt : moment(0).valueOf()
    },
    {
        id: '2',
        description: 'merda2',
        amount : 15,
        note: 'merdamerda2',
        createdAt : moment(0).subtract(4, 'days').valueOf()
    },
    {
        id: '3',
        description: 'merda3',
        amount : 300,
        note: 'merdamerda3',
        createdAt : moment(0).add(4, 'days').valueOf()
    }
]

test('testo la lista di item verso snapshot', () => {
    const wrapper = shallow(<ExpenseList expenses={testData} />);

    expect(wrapper).toMatchSnapshot(); 
});


test('test should render expense list with empty message', () => {
    const wrapper = shallow(<ExpenseList expenses={[]} />);
    
    expect(wrapper).toMatchSnapshot();
})