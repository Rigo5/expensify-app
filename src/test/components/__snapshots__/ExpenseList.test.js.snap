// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test should render expense list with empty message 1`] = `
<div>
  <p>
    No expenses
  </p>
</div>
`;

exports[`testo la lista di item verso snapshot 1`] = `
<div>
  <Connect(ExpensesListItem)
    amount={10}
    createdAt={0}
    description="merda"
    id="1"
    key="1"
    note="merdamerda"
  />
  <Connect(ExpensesListItem)
    amount={15}
    createdAt={-345600000}
    description="merda2"
    id="2"
    key="2"
    note="merdamerda2"
  />
  <Connect(ExpensesListItem)
    amount={300}
    createdAt={345600000}
    description="merda3"
    id="3"
    key="3"
    note="merdamerda3"
  />
</div>
`;
